### ISSUE TITLE
Implement new feature CopyCurrentState

### ISSUE THREAD

**Issue by alemhg** [fortify/fcli#460]
Hello,

I'm working developing a new feature to copyCurrentState for Fortify SSC application versions (Gradle + Java 17).

The command is working, after executing it the source application version is copied to the target application version

```
{
    "data" : {
        "status" : "success",
        "message" : "Copy current state action succeeded",
        "values" : null
    },
 "responseCode" : 200
}
```
but I get the following error

```
org.springframework.expression.spel.SpelEvaluationException: EL1007E: Property or field 'name' cannot be found on null
 at org.springframework.expression.spel.ast.PropertyOrFieldReference.readProperty(PropertyOrFieldReference.java:218)
 at org.springframework.expression.spel.ast.PropertyOrFieldReference.getValueInternal(PropertyOrFieldReference.java:105)
 at org.springframework.expression.spel.ast.PropertyOrFieldReference$AccessorLValue.getValue(PropertyOrFieldReference.java:410)
 at org.springframework.expression.spel.ast.CompoundExpression.getValueInternal(CompoundExpression.java:96)
 at org.springframework.expression.spel.ast.SpelNodeImpl.getTypedValue(SpelNodeImpl.java:119)
 at org.springframework.expression.spel.standard.SpelExpression.getValue(SpelExpression.java:376)
 at com.fortify.cli.common.spring.expression.SpelEvaluator.evaluate(SpelEvaluator.java:46)
 at com.fortify.cli.common.spring.expression.SpelEvaluator.evaluate(SpelEvaluator.java:50)
 at com.fortify.cli.common.json.JsonHelper.evaluateSpelExpression(JsonHelper.java:64)
 at com.fortify.cli.common.output.writer.record.AbstractFormattedRecordWriter.evaluateValue(AbstractFormattedRecordWriter.java:93)
 at com.fortify.cli.common.output.writer.record.AbstractFormattedRecordWriter.lambda$applyOptionalFieldPathsTransformation$0(AbstractFormattedRecordWriter.java:81)
 at java.base/java.util.Arrays$ArrayList.forEach(Arrays.java:4204)
 at com.fortify.cli.common.output.writer.record.AbstractFormattedRecordWriter.applyOptionalFieldPathsTransformation(AbstractFormattedRecordWriter.java:80)
 at com.fortify.cli.common.output.writer.record.AbstractFormattedRecordWriter.getFormattedRecord(AbstractFormattedRecordWriter.java:74)
 at com.fortify.cli.common.output.writer.record.AbstractFormattedRecordWriter.writeRecord(AbstractFormattedRecordWriter.java:53)
 at com.fortify.cli.common.output.writer.output.standard.StandardOutputWriter$AbstractRecordWriterWrapper.writeRecord(StandardOutputWriter.java:298)
 at com.fortify.cli.common.output.writer.output.standard.StandardOutputWriter$OutputAndVariableRecordWriter.writeRecord(StandardOutputWriter.java:266)
 at com.fortify.cli.common.output.writer.output.standard.StandardOutputWriter.writeRecord(StandardOutputWriter.java:195)
 at com.fortify.cli.common.output.writer.output.standard.StandardOutputWriter.writeRecords(StandardOutputWriter.java:170)
 at com.fortify.cli.common.output.writer.output.standard.StandardOutputWriter.writeRecords(StandardOutputWriter.java:151)
 at com.fortify.cli.common.output.writer.output.standard.StandardOutputWriter.lambda$writeRecords$1(StandardOutputWriter.java:140)
 at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
 at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
 at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
 at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
 at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
 at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
 at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
 at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
 at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
 at kong.unirest.PagedList.ifSuccess(PagedList.java:52)
 at com.fortify.cli.common.output.writer.output.standard.StandardOutputWriter.writeRecords(StandardOutputWriter.java:140)
 at com.fortify.cli.common.output.writer.output.standard.StandardOutputWriter.write(StandardOutputWriter.java:103)
 at com.fortify.cli.common.output.cli.mixin.AbstractOutputHelperMixin.write(AbstractOutputHelperMixin.java:68)
 at com.fortify.cli.common.output.cli.cmd.AbstractOutputCommand.run(AbstractOutputCommand.java:31)
 at picocli.CommandLine.executeUserObject(CommandLine.java:2103)
 at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2538)
 at picocli.CommandLine$RunLast.handle(CommandLine.java:2530)
 at picocli.CommandLine$RunLast.handle(CommandLine.java:2492)
 at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2350)
 at picocli.CommandLine$RunLast.execute(CommandLine.java:2494)
 at picocli.CommandLine.execute(CommandLine.java:2247)
 at com.fortify.cli.app.runner.DefaultFortifyCLIRunner.run(DefaultFortifyCLIRunner.java:49)
 at com.fortify.cli.app.FortifyCLI.execute(FortifyCLI.java:38)
 at com.fortify.cli.app.FortifyCLI.main(FortifyCLI.java:32)
```

This is my command

```
package com.fortify.cli.ssc.appversion.cli.cmd;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fortify.cli.common.output.cli.mixin.IOutputHelper;
import com.fortify.cli.common.output.cli.mixin.OutputHelperMixins;
import com.fortify.cli.common.output.product.IProductHelper;
import com.fortify.cli.common.output.transform.IActionCommandResultSupplier;
import com.fortify.cli.common.output.transform.IRecordTransformer;
import com.fortify.cli.ssc._common.output.cli.cmd.AbstractSSCJsonNodeOutputCommand;
import com.fortify.cli.ssc._common.output.cli.cmd.AbstractSSCBaseRequestOutputCommand;
import com.fortify.cli.ssc._common.rest.SSCUrls;
import com.fortify.cli.ssc._common.rest.bulk.SSCBulkRequestBuilder;
import com.fortify.cli.ssc._common.rest.bulk.SSCBulkRequestBuilder.SSCBulkResponse;
import com.fortify.cli.ssc.appversion.cli.mixin.SSCAppVersionResolverMixin;
import com.fortify.cli.ssc.appversion.helper.SSCAppVersionDescriptor;
import com.fortify.cli.ssc.appversion.helper.SSCAppVersionHelper;
import com.fortify.cli.ssc.appversion_attribute.helper.SSCAppVersionAttributeUpdateBuilder;
import com.fortify.cli.ssc.appversion_user.helper.SSCAppVersionUserUpdateBuilder;
import kong.unirest.HttpRequest;
import kong.unirest.RequestBodyEntity;
import kong.unirest.UnirestInstance;
import lombok.Getter;
import picocli.CommandLine.Command;
import picocli.CommandLine.Mixin;
import picocli.CommandLine.Option;

//OutputHelperMixins.CopyCurrentState.CMD_NAME = "copy-state"
@Command(name = OutputHelperMixins.CopyCurrentState.CMD_NAME)
public class SSCAppVersionCopyCurrentStateCommand extends AbstractSSCBaseRequestOutputCommand
		implements IActionCommandResultSupplier {
	@Getter
	@Mixin
	private OutputHelperMixins.CopyCurrentState outputHelper;
	@Mixin
	private SSCAppVersionResolverMixin.PositionalParameter sourceAppVersionResolver;

	@Option(names = { "--target-appversion", "-t" }, required = true)
	private String targetAppVersion;

	@Override
	public boolean isSingular() {
		System.out.println("copy-state: isSingular: true");
		return true;
	}

	@Override
	public String getActionCommandResult() {
		System.out.println("copy-state: getActionCommandResult: COPIED_CURRENT_STATE");
		return "COPIED_CURRENT_STATE";
	}

	@Override
	public HttpRequest<?> getBaseRequest(UnirestInstance unirest) {
		System.out.println("getJsonNodecopy-state: getJsonNode");
		SSCAppVersionDescriptor descriptorSourceAppVersion = sourceAppVersionResolver.getAppVersionDescriptor(unirest);
		String sourceAppVersion = descriptorSourceAppVersion.getVersionId();
		System.out.println("getJsonNodecopy-state: getJsonNode: sourceAppVersion " + sourceAppVersion);
		System.out.println("getJsonNodecopy-state: getJsonNode: targetAppVersion " + this.targetAppVersion);
		ObjectNode data = (new ObjectMapper()).createObjectNode();
		data.put("type", "COPY_CURRENT_STATE");
		ObjectNode values = (new ObjectMapper()).createObjectNode();
		values.put("projectVersionId", Integer.valueOf(this.targetAppVersion));
		values.put("previousProjectVersionId", Integer.valueOf(sourceAppVersion));
		data.set("values", values);
		System.out.println("getJsonNodecopy-state: getJsonNode: post body: " + data.toString());
		return unirest.post("/api/v1/projectVersions/" + this.targetAppVersion + "/action").body(data);
	}
}


OutputHelperMixins.CopyCurrentState.CMD_NAME = "copy-state"

```

Do you have any clue/hint about this issue?

Best regards,

Alejo

[SSCAppVersionCopyCurrentStateCommand.java.txt](https://github.com/fortify/fcli/files/12916053/SSCAppVersionCopyCurrentStateCommand.java.txt)
[OutputHelperMixins.java.txt](https://github.com/fortify/fcli/files/12916054/OutputHelperMixins.java.txt)
[SSCAppVersionCommands.java.txt](https://github.com/fortify/fcli/files/12916055/SSCAppVersionCommands.java.txt)
[SSCMessages.properties.txt](https://github.com/fortify/fcli/files/12916211/SSCMessages.properties.txt)

**Comment by rsenden** [comment]
Hi, this exception is caused by the following resource bundle entry in SSCMessages.properties (assuming you're working off the `develop` branch, might be slightly different on `main` branch which currently is only meant for 1.x bug fixes):

```
fcli.ssc.appversion.output.table.options = id,application.name,name,issueTemplateName,createdBy
```

All `appversion` sub-commands are expected to return an application version record; the endpoint you're invoking returns just a `success` response and doesn't include any of the properties listed in the default table output properties as shown above.  For testing, you can for example specify `-o json` to avoid these default table output properties from being used.

Most `appversion` sub-commands retrieve the appropriate `SSCAppVersionDescriptor`, either through resolver mixin or `SSCAppVersionHelper`, and then return `descriptor.asJsonNode()` or `descriptor.asObjectNode()` (with command class extending from `AbstractSSCJsonNodeOutputCommand` instead of `AbstractSSCBaseRequestOutputCommand`).

Some notes:

* Usually, copying state is only done when creating a new application version, so it may be more logical to have this as an option on the `appversion create` command (somewhat similar to the `fcli fod release create --copy-from` option) unless you think there's a proper use case for copying state after an application version has been created.
* Unless you're an OpenText employee, we'll need a signed contribution agreement to be submitted if you'd want us to add contributions like these to fcli; see https://github.com/fortify/fcli/blob/develop/CONTRIBUTING.md.

**Comment by alemhg** [comment]
Hi @rsenden,

Thanks a lot for your feedback and proposed solution.

For sure, we will provide the signed contribution agreement before starting any pull request.

Best regards,
Alejo

**Comment by rsenden** [comment]
This issue seems no longer relevant; full state copy functionality will be provided on the `fcli ssc appversion create` command in the next fcli release, and a fix has been implemented to no longer throw an exception if a given property cannot be found.
